hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/TokenUpdateLogic.java

```diff
@@ -0,0 +1,246 @@
+import java.util.Optional;
+import javax.inject.Inject;
+
+public class TokenUpdateLogic {
+    private final OptionValidator validator;
+    private final HederaTokenStore tokenStore;
+    private final WorldLedgers worldLedgers;
+    private final SideEffectsTracker sideEffectsTracker;
+    private final SigImpactHistorian sigImpactHistorian;
+    boolean allowChangedTreasuryToOwnNfts;
+
+    @Inject
+    public TokenUpdateLogic(
+            final @AreTreasuryWildcardsEnabled boolean allowChangedTreasuryToOwnNfts,
+            OptionValidator validator,
+            HederaTokenStore tokenStore,
+            WorldLedgers worldLedgers,
+            SideEffectsTracker sideEffectsTracker,
+            SigImpactHistorian sigImpactHistorian) {
+        this.validator = validator;
+        this.tokenStore = tokenStore;
+        this.worldLedgers = worldLedgers;
+        this.sideEffectsTracker = sideEffectsTracker;
+        this.sigImpactHistorian = sigImpactHistorian;
+        this.allowChangedTreasuryToOwnNfts = allowChangedTreasuryToOwnNfts;
+    }
+
+    public void updateToken(TokenUpdateTransactionBody op, long now) {
+        final var tokenID = Id.fromGrpcToken(op.getToken()).asGrpcToken();
+        validateFalse(tokenID.equals(MISSING_TOKEN), INVALID_TOKEN_ID);
+        if (op.hasExpiry()) {
+            validateTrueOrRevert(validator.isValidExpiry(op.getExpiry()), INVALID_EXPIRATION_TIME);
+        }
+        MerkleToken token = tokenStore.get(tokenID);
+        checkTokenPreconditions(token, op);
+
+        ResponseCodeEnum outcome = autoRenewAttachmentCheck(op, token);
+        validateTrueOrRevert(outcome == OK, outcome);
+
+        Optional<AccountID> replacedTreasury = Optional.empty();
+        if (op.hasTreasury()) {
+            var newTreasury = op.getTreasury();
+            validateFalseOrRevert(isDetached(newTreasury), ACCOUNT_EXPIRED_AND_PENDING_REMOVAL);
+
+            if (!tokenStore.associationExists(newTreasury, tokenID)) {
+                outcome = tokenStore.autoAssociate(newTreasury, tokenID);
+                if (outcome != OK) {
+                    abortWith(outcome);
+                }
+            }
+            var existingTreasury = token.treasury().toGrpcAccountId();
+            if (!allowChangedTreasuryToOwnNfts && token.tokenType() == NON_FUNGIBLE_UNIQUE) {
+                var existingTreasuryBalance = getTokenBalance(existingTreasury, tokenID);
+                if (existingTreasuryBalance > 0L) {
+                    abortWith(CURRENT_TREASURY_STILL_OWNS_NFTS);
+                }
+            }
+            if (!newTreasury.equals(existingTreasury)) {
+                validateFalseOrRevert(
+                        isDetached(existingTreasury), ACCOUNT_EXPIRED_AND_PENDING_REMOVAL);
+
+                outcome = prepTreasuryChange(tokenID, token, newTreasury, existingTreasury);
+                if (outcome != OK) {
+                    abortWith(outcome);
```

Comment:
All the `abortWith` cases are missing  to have test coverage

Reply:
This is a strange sonarcloud behaviour. I checked the unit tests and while debugging I can see the code execution goes through all `abortWith` methods. However the sonar does not check these lines as covered.